lmdeploy/pytorch/paging/block_trie.py

```diff
@@ -0,0 +1,215 @@
+        return True
+
+
+class BlockTrie:
+
+    def __init__(self, cache_config: CacheConfig,
+                 block_manager: BaseBlockManager):
+        self.block_manager = block_manager
+        self.cache_config = cache_config
+        self.allocator = self.block_manager.allocator
+        self.block_size = cache_config.block_size
+        self.enable = self.cache_config.prefix_caching
+
+        # caches with different adapter should not be shared.
+        self._roots: Dict[str, Node] = dict()
+        self.leaves: Set[Node] = set()
+
+    def get_root(self, adapter_name: str):
+        if adapter_name not in self._roots:
+            self._roots[adapter_name] = Node(-1, -1, None)
+        return self._roots[adapter_name]
+
+    def match(self, seq: SchedulerSequence):
+        if not self.enable:
+            return
+
+        block_size = self.block_size
+        matched_blocks = []
+
+        logical_blocks = seq.logical_blocks
+        curr: Node = getattr(logical_blocks, 'last_shared_node', None)
+        if curr is None:
+            curr = self.get_root(seq.adapter_name)
+        num_matched = curr.num_matched
+
+        def __match_success(node: Node):
+            nonlocal curr, num_matched
+            matched_blocks.append(node.block)
+            curr = node
+            num_matched += block_size
+
+        while num_matched + block_size < seq.num_all_ids:
+            curr_tokens = seq.history_cache[num_matched:num_matched +
+                                            block_size]
+
+            key = hash(tuple(curr_tokens))
+            if key not in curr.children:
+                break
+
+            child = curr.children[key]
+            if not np.array_equal(curr_tokens, child.tokens):
+                break
+
+            __match_success(child)
+
+        if len(matched_blocks) > 0:
+            matched_blocks = np.array(matched_blocks)
+            self.allocator.update_access_time(matched_blocks)
+            self.allocator.add_ref_count(matched_blocks, 1)
+            seq.logical_blocks.append(matched_blocks)
+            seq.set_step(num_matched)
+
+        seq.logical_blocks.last_shared_node = curr
+
+    def allocate(self, seq: SchedulerSequence, prealloc_size: int = 0):
+        self.block_manager.allocate(seq, prealloc_size)
+
+        if not self.enable:
+            return
+
+        block_size = self.block_size
+        logical_blocks = seq.logical_blocks
+        node: Node = getattr(logical_blocks, 'last_shared_node', None)
+        if node is None:
+            node = self.get_root(seq.adapter_name)
+
+        num_matched = node.num_matched
+        num_all_ids = seq.num_all_ids
+
+        if num_matched + block_size > num_all_ids:
+            return
+
+        if len(node.children) == 0 and node.parent is not None:
+            self.leaves.remove(node)
+
+        block_id = num_matched // block_size
+        blocks = []
+        free_blocks = []
+        while num_matched + block_size < num_all_ids:
```

Comment:
Maybe we can consider `<=` here. For example, if `num_all_ids=128`, `num_matched=64`, `block_size=64`, the second block can also be cached.

Reply:
all_ids == token_ids + history_ids, cache all ids might break engine main loop.

Reply:
Do you mean here `all_ids` contains some blocks that have not been computed?

Reply:
yes

Reply:
Got it. Why not filter this `==` case in `L123`?

Reply:
We only need to leave the last token id to make the engine work, other token ids can be cached. This filter should be enough.

Reply:
I am not sure whether you mention the `matching` or `inserting`. In `matching`, we do need to leave at least one token left to make the engine work. But in `inserting`, I am not sure if we also need to consider this.

Reply:
I see, `<=` is fine.